lib/cdo/aws/rds.rb

```diff
@@ -0,0 +1,120 @@
+require_relative '../../../deployment'
+require 'aws-sdk-rds'
+
+module Cdo
+  class RDS
+    # Create an RDS Aurora MySQL cluster cloned from a specified source cluster.
+    # @param source_cluster_id [String] DB cluster id of the cluster to clone.  Defaults to current environment's cluster.
+    # @param clone_cluster_id [String] DB cluster id to assign to clone.  Defaults to source cluster id + "-clone"
+    # @param instance_type [String]
+    def self.clone_cluster(
+      source_cluster_id: CDO.db_cluster_id,
+      clone_cluster_id: "#{source_cluster_id}-clone",
+      instance_type: 'db.r4.large'
+    )
+      clone_instance_id = clone_cluster_id + "-0"
+      rds_client = Aws::RDS::Client.new
+      begin
+        CDO.log.info "Creating clone of database cluster - #{source_cluster_id}"
+        source_cluster = rds_client.describe_db_clusters({db_cluster_identifier: source_cluster_id}).db_clusters.first
+        rds_client.restore_db_cluster_to_point_in_time(
+          {
+            db_cluster_identifier: clone_cluster_id,
+            restore_type: 'copy-on-write',
+            source_db_cluster_identifier: source_cluster_id,
+            use_latest_restorable_time: true,
+            db_subnet_group_name: source_cluster.db_subnet_group,
+            vpc_security_group_ids: source_cluster.vpc_security_groups.map(&:vpc_security_group_id),
+          }
+        )
+        source_writer_instance_identifier = source_cluster.
+          db_cluster_members.
+          select(&:is_cluster_writer).
+          first.
+          db_instance_identifier
+        source_writer_instance = rds_client.
+          describe_db_instances({db_instance_identifier: source_writer_instance_identifier}).
+          db_instances.
+          first
+        rds_client.create_db_instance(
+          {
+            db_instance_identifier: clone_instance_id,
+            db_instance_class: instance_type,
+            engine: source_cluster.engine,
+            db_cluster_identifier: clone_cluster_id,
+            db_parameter_group_name: source_writer_instance.db_parameter_groups[0].db_parameter_group_name
```

Comment:
Should we duplicate the source cluster/instance Parameter Groups instead of re-using them?  What's the risk of someone cloning the production cluster and then inadvertently modifying the production cluster when they experiment with changes to their clone's parameter group?

Reply:
I went with Option 2 because it ensures that engineer cloning a cluster gets a new cluster that is as similar as possible to the one they're cloning without risking that they'll inadvertently change settings on the source cluster.